// geoman-client/build/classes/District.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var Subdistrict_1 = require("./Subdistrict");
var turf = __importStar(require("@turf/turf"));
/**
 * Class District yang menyimpan data kecamatan
 * @class District
 */
var District = /** @class */ (function () {
    /**
     * Membuat instance dari class District
     * @param geoman instance class GeoMan
     * @param rawData data json raw dari backend
     */
    function District(geoman, rawData) {
        this.id = rawData.id;
        this.name = rawData.name;
        this.properties = rawData.properties;
        this.created_at = rawData.created_at;
        this.updated_at = rawData.updated_at;
        this.bbox = rawData.bbox;
        this.geoman = geoman;
    }
    /**
     * Arahkan map ke kecamatan ini
     */
    District.prototype.focus = function () {
        var _this = this;
        this.geoman.map.fitBounds(this.bbox);
        if (this.geoman.map.getLayer('region-lay')) {
            this.geoman.map.removeLayer('region-lay');
            this.geoman.map.removeSource('region-lay');
        }
        return this.getShape().then(function (shape) {
            var earth = turf.polygon([[[-180, -90], [-180, 90], [180, 90], [180, -90], [-180, -90]]]);
            var lay = turf.difference(earth, shape);
            _this.geoman.map.addLayer({
                id: 'region-lay',
                type: 'fill',
                source: {
                    type: 'geojson',
                    data: lay
                },
                paint: {
                    'fill-color': 'rgba(0,0,0,.6)'
                }
            });
        });
    };
    /**
     * Ambil daftar kelurahan dibawah kecamatan ini
     */
    District.prototype.getSubdistricts = function () {
        var _this = this;
        return this.geoman.http.get("/maps/districts/" + this.id + "/subdistricts")
            .then(function (data) {
            return data.map(function (r) { return new Subdistrict_1.Subdistrict(_this.geoman, _this, r); });
        });
    };
    /**
     * Ambil bentuk (geometry) dari kecamatan ini
     */
    District.prototype.getShape = function () {
        return this.geoman.http.get("/maps/districts/" + this.id)
            .then(function (data) { return data.geometry; });
    };
    return District;
}());
exports.District = District;
