// geoman-client/build/classes/Neighbor.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var turf = __importStar(require("@turf/turf"));
/**
 * Class Neighbor yang menyimpan data kecamatan
 * @class Neighbor
 */
var Neighbor = /** @class */ (function () {
    /**
     * Membuat instance dari class Neighbor
     * @param geoman instance class GeoMan
     * @param rawData data json raw dari backend
     */
    function Neighbor(geoman, district, subdistrict, rawData) {
        this.id = rawData.id;
        this.name = rawData.name;
        this.properties = rawData.properties;
        this.created_at = rawData.created_at;
        this.updated_at = rawData.updated_at;
        this.district_id = rawData.district_id;
        this.subdistrict_id = rawData.subdistrict_id;
        this.bbox = rawData.bbox;
        this.district = district;
        this.subdistrict = subdistrict;
        this.geoman = geoman;
    }
    /**
     * Arahkan map ke lingkungan ini
     */
    Neighbor.prototype.focus = function () {
        var _this = this;
        this.geoman.map.fitBounds(this.bbox);
        if (this.geoman.map.getLayer('region-lay')) {
            this.geoman.map.removeLayer('region-lay');
            this.geoman.map.removeSource('region-lay');
        }
        return this.getShape().then(function (shape) {
            var earth = turf.polygon([[[-180, -90], [-180, 90], [180, 90], [180, -90], [-180, -90]]]);
            var lay = turf.difference(earth, shape);
            _this.geoman.map.addLayer({
                id: 'region-lay',
                type: 'fill',
                source: {
                    type: 'geojson',
                    data: lay
                },
                paint: {
                    'fill-color': 'rgba(0,0,0,.6)'
                }
            });
        });
    };
    /**
     * Ambil bentuk (geometry) dari kelurahan ini
     */
    Neighbor.prototype.getShape = function () {
        return this.geoman.http.get("/maps/districts/" + this.district.id + "/subdistricts/" + this.subdistrict.id + "/neighbors/" + this.id)
            .then(function (data) { return data.geometry; });
    };
    return Neighbor;
}());
exports.Neighbor = Neighbor;
